data_model/src/events/data/filters.rs

```diff
@@ -1,192 +1,466 @@
+    #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    pub struct PeerEventFilter {
+        pub id_matcher: Option<super::PeerId>,
+        pub event_matcher: Option<PeerEventMatcher>,
     }
 
-    /// Filter that accepts a data event with the matching origin.
-    #[derive(
-        Clone,
-        PartialOrd,
-        Ord,
-        Eq,
-        Debug,
-        Constructor,
-        Decode,
-        Encode,
-        Serialize,
-        Deserialize,
-        IntoSchema,
-    )]
-    #[serde(bound(
-        deserialize = "<<T as HasOrigin>::Origin as Identifiable>::Id: Deserialize<'de>"
-    ))]
-    #[serde(transparent)]
-    #[repr(transparent)]
-    pub struct OriginFilter<T: HasOrigin>(pub(super) <T::Origin as Identifiable>::Id)
-    where
-        <T::Origin as Identifiable>::Id:
-            Debug + Clone + Eq + Ord + Decode + Encode + Serialize + IntoSchema;
+    /// An event matcher for [`PeerEvent`]s
+    #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    pub enum PeerEventMatcher {
+        /// Matches only [`PeerEvent::Added`]
+        ByAdded,
+        /// Matches only [`PeerEvent::Removed`]
+        ByRemoved,
+    }
+
+    /// An event filter for [`DomainEvent`]s
+    #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    pub struct DomainEventFilter {
+        /// If specified matches only events originating from this domain
+        pub id_matcher: Option<super::DomainId>,
+        /// If specified matches only events of this type
+        pub event_matcher: Option<DomainEventMatcher>,
+    }
+
+    /// An event matcher for [`DomainEvent`]s
+    #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    pub enum DomainEventMatcher {
+        /// Matches only [`DomainEvent::Created`]
+        ByCreated,
+        /// Matches only [`DomainEvent::Deleted`]
+        ByDeleted,
+        /// Matches only [`DomainEvent::MetadataInserted`]
+        ByMetadataInserted,
+        /// Matches only [`DomainEvent::MetadataRemoved`]
+        ByMetadataRemoved,
+        /// Matches only [`DomainEvent::OwnerChanged`]
+        ByOwnerChanged,
+        // we allow filtering for nested events, but if you need to specify an id matcher for, for example, AccountId, you need to use AccountFilter
+        /// Matches only [`DomainEvent::Account`]
+        ByAccountAny,
+        /// Matches only [`DomainEvent::AssetDefinition`]
+        ByAssetDefinitionAny,
+    }
+
+    /// An event filter for [`AccountEvent`]s
+    #[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+    pub struct AccountEventFilter {
+        /// If specified matches only events originating from this account
+        pub id_matcher: Option<super::AccountId>,
+        /// If specified matches only events of this type
+        pub event_matcher: Option<AccountEventMatcher>,
+    }
```

Comment:
```suggestion
    pub struct AccountEventFilter {
        /// If specified matches only events originating from this account
        pub id_matcher: Option<super::AccountId>,
        /// If specified matches only events of this type
        pub event_matcher: AccountEventMatcher,
    }
```

now instead of having `Option` here I would rather have `ByAny` variant in `AccountEventMatcher` like you have for `NotificationEventFilter`

Reply:
For some reason I thought I did this already... Yes, I think it's a good idea!

Reply:
Not applicable anymore with multi-event matchers